src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php

```diff
@@ -0,0 +1,345 @@
+
+	/**
+	 * Constructs the controller.
+	 */
+	public function __construct() {
+		$this->namespace = 'wp/v2';
+		$this->rest_base = 'block-directory';
+	}
+
+	/**
+	 * Registers the necessary REST API routes.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/search',
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_items' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+					'args'                => $this->get_collection_params(),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Checks whether a given request has permission to install and activate plugins.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|bool True if the request has permission, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) {
+		if ( ! current_user_can( 'install_plugins' ) || ! current_user_can( 'activate_plugins' ) ) {
+			return new WP_Error(
+				'rest_block_directory_cannot_view',
+				__( 'Sorry, you are not allowed to browse the block directory.' ),
+				array( 'status' => rest_authorization_required_code() )
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Search and retrieve blocks metadata
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
+		require_once ABSPATH . 'wp-admin/includes/plugin-install.php';
+		require_once ABSPATH . 'wp-admin/includes/plugin.php';
+
+		$response = plugins_api(
+			'query_plugins',
+			array(
+				'block'    => $request['term'],
+				'per_page' => $request['per_page'],
+				'page'     => $request['page'],
+			)
+		);
+
+		if ( is_wp_error( $response ) ) {
+			$response->add_data( array( 'status' => 500 ) );
+
+			return $response;
+		}
+
+		$result = array();
+
+		foreach ( $response->plugins as $plugin ) {
+			if ( $this->find_plugin_for_slug( $plugin['slug'] ) ) {
+				continue;
+			}
+
+			$data     = $this->prepare_item_for_response( $plugin, $request );
+			$result[] = $this->prepare_response_for_collection( $data );
+		}
+
+		return rest_ensure_response( $result );
+	}
+
+	/**
+	 * Parse block metadata for a block, and prepare it for an API repsonse.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param array           $plugin  The plugin metadata.
+	 * @param WP_REST_Request $request Request object.
+	 *
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function prepare_item_for_response( $plugin, $request ) {
+		// There might be multiple blocks in a plugin. Only the first block is mapped.
+		$block_data = reset( $plugin['blocks'] );
```

Comment:
Probably a question for @tellyworth — How will we ensure the right block is imported here? I know [Listicles](https://wordpress.org/plugins/listicles/) is an example of this, it pulls out the List Item child block instead of the main Listicles block. The wp.org endpoint could return `parent`, and we filter out based on the current block, or maybe it only returns blocks with no parent?